Add scanpy chat assistant and documentation processing scripts

ahuang11 opened the pull request:
Adds:
1. app for a Chat with scanpy expert
2. convert_docs (html to markdown)
3. embed_docs (add to vector store)
4. scrape_docs (download all the HTMLs from docs)

Put here for now until we discuss how we want to integrate (whether Lumen / Lumen-Anndata) and what to do with the embeddings scanpy.db with a size of 9.8M.

![image](https://github.com/user-attachments/assets/b74ed003-8215-43f1-a5c2-f81fc3eca3fa)

droumis commented:
How to handle versioning? Would be really nice to incorporate the version into the embeddings

droumis commented:
We decided at the meeting to keep things in this repo for now, until a Lumen core generalization is further debated

droumis commented:
Regarding where to put the docs embeddings.... IMO:

For the initial implementation, let's bundle the current docs version embeddings (9.8MB) with the package - this is a reasonable size that won't significantly impact installation. This gets us immediate functionality and a quick development path to get other features built and tested. 

Then later, when support for multiple versions becomes a higher priority, we can implement a version checking mechanism and add a cloud storage pull for additional versions (HoloViz S3, etc) with a simple caching system that triggers on app launch (or something).

@philippjfr , any thoughts? I guess this all also depends on lumen-anndata becoming a package, which I'm still assuming it should.

ahuang11 commented:
I think this PR can be simplified now--removing the download and convert scripts.

Rather than manually scraping the HTML and converting to markdown, I tried building it natively using https://github.com/liran-funaro/sphinx-markdown-builder and it seems to work well at a glance for scanpy.

The downside is that users need to
[add the markdown builder extension](https://github.com/liran-funaro/sphinx-markdown-builder?tab=readme-ov-file#usage) and requirement
install all the existing package's docs dependencies
know how to build the docs...

droumis commented:
Why do users need to build the docs if we deliver the embeddings (or access to them) as part of a release?

ahuang11 commented:
Okay I moved things around and added the updated database with versioning.

I migrated most of the functionality into Lumen:
https://github.com/holoviz/lumen/pull/1221

Here's an example generated prompt:
```python
You are an expert scientist working in Python, with a specialty using Anndata and Scanpy.
Help the user with their questions, and if you don&#39;t know the answer, say so.

Here was the plan that was executed:
"""
The user is asking for the structure of AnnData, which is a data structure used in single-cell genomics. This question is high-level and focuses on understanding the components and organization of AnnData. The relevant actors for this query are likely to be ChatAgent for discussing the structure and features of AnnData, as well as potentially TableLookup for any related data structure details. However, since I already have sufficient information about AnnData's structure in memory, I can proceed with using ChatAgent directly without needing to look up additional tables.
"""

Below are the relevant tables and columns available:






Context from tools:
"""

Please augment your response with the following context if relevant:
- At the most basic level, an [`AnnData`](https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData) object `adata` stores
a data matrix `adata.X`, annotation of observations
`adata.obs` and variables `adata.var` as `pd.DataFrame` and unstructured
annotation `adata.uns` as `dict`. Names of observations and
variables can be accessed via `adata.obs_names` and `adata.var_names`,
respectively. [`AnnData`](https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData) objects can be sliced like
dataframes, for example, `adata_subset = adata[:, list_of_gene_names]`.
For more, see this [blog post](https://falexwolf.de/blog/171223_AnnData_indexing_views_HDF5-backing/).

To read a data file to an [`AnnData`](https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData) object, call:

```default
adata = sc.read(filename)
```

to initialize an [`AnnData`](https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData) object. Possibly add further annotation using, e.g., `pd.read_csv`:

```default
import pandas as pd
anno = pd.read_csv(filename_sample_annotation)
adata.obs['cell_groups'] = anno['cell_groups']  # categorical annotation of type pandas.Categorical
adata.obs['time'] = anno['time']                # numerical annotation of type float
# alternatively, you could also set the whole dataframe
# adata.obs = anno
```

To write, use:

```default
adata.write_h5ad(filename)
adata.write_zarr(filename)
adata.write_csvs(filename)
adata.write_loom(filename)
```
 (Relevance: 0.6 - Metadata: {'version': '1.11.1', 'filename': 'scanpy_1.11.1/usage-principles.md', 'llm_context': 'Scanpy docs: AnnData object structure, data reading/writing methods, and annotation handling.'})
-   In “denoise” mode, `adata.X` is overwritten with the denoised values.
  In “latent” mode, latent low dimensional representation of cells are stored
  in `adata.obsm['X_dca']` and `adata.X` is not modified.
  Note that these values are not corrected for library size effects.

  If `return_info` is true, all estimated distribution parameters are stored
  in AnnData like this:

  `.obsm["X_dca_dropout"]`
  : The mixture coefficient (pi) of the zero component in ZINB,
    i.e. dropout probability (if `ae_type` is `zinb` or `zinb-conddisp`).

  `.obsm["X_dca_dispersion"]`
  : The dispersion parameter of NB.

  `.uns["dca_loss_history"]`
  : The loss history of the training.
    See `.history` attribute of Keras History class for mode details.

  Finally, the raw counts are stored in `.raw` attribute of AnnData object.

  If `return_model` is given, trained model is returned.
  When both `copy` and `return_model` are true,
  a tuple of anndata and model is returned in that order.
 (Relevance: 0.6 - Metadata: {'version': '1.11.1', 'filename': 'scanpy_1.11.1/generated/scanpy.external.pp.dca.md', 'llm_context': 'Scanpy docs: DCA return values in denoise/latent modes, including dropout, dispersion parameters, and loss history.'})
- If your anndata object has a `"spatial"` entry in `.uns`, the `img_key`
and `library_id` parameters to find values for `img`, `scale_factor`,
and `spot_size` arguments. Alternatively, these values be passed directly.

* **Parameters:**
  **adata**
  : Annotated data matrix.

  **color**
  : Keys for annotations of observations/cells or variables/genes, e.g.,
    `'ann1'` or `['ann1', 'ann2']`.

  **gene_symbols**
  : Column name in `.var` DataFrame that stores gene symbols. By default `var_names`
    refer to the index column of the `.var` DataFrame. Setting this option allows
    alternative names to be used.

  **use_raw**
  : Use `.raw` attribute of `adata` for coloring with gene expression. If `None`,
    defaults to `True` if `layer` isn’t provided and `adata.raw` is present.

  **layer**
  : Name of the AnnData object layer that wants to be plotted. By default
    adata.raw.X is plotted. If `use_raw=False` is set, then `adata.X` is plotted.
    If `layer` is set to a valid layer name, then the layer is plotted. `layer`
    takes precedence over `use_raw`.

  **library_id**
  : library_id for Visium data, e.g. key in `adata.uns["spatial"]`.

  **img_key**
  : Key for image data, used to get `img` and `scale_factor` from `"images"`
    and `"scalefactors"` entires for this library. To use spatial coordinates,
    but not plot an image, pass `img_key=None`.

  **img**
  : image data to plot, overrides `img_key`.

  **scale_factor**
  : Scaling factor used to map from coordinate space to pixel space.
    Found by default if `library_id` and `img_key` can be resolved.
    Otherwise defaults to `1.`.

  **spot_size**
  : Diameter of spot (in coordinate space) for each point. Diameter
    in pixels of the spots will be `size * spot_size * scale_factor`.
    This argument is required if it cannot be resolved from library info. (Relevance: 0.6 - Metadata: {'version': '1.11.1', 'filename': 'scanpy_1.11.1/api/generated/scanpy.pl.spatial.md', 'llm_context': 'Scanpy API ref: Parameters for `scanpy.pl.spatial()`, including `adata`, `color`, `layer`, and image settings.'})
"""
```

droumis commented:
To capture the install instructions, which are different than what is currently in the README:

```bash
git clone https://github.com/holoviz-topics/lumen-anndata.git
cd lumen-anndata
git switch add_scripts

python3 -m venv .venv

pip install "git+https://github.com/holoviz/lumen@main#egg=lumen[ai]"
pip install -e .

cd src/lumen_anndata/

panel serve app.py
```

droumis reviewed (changes requested):
Looks good so far. Just please ensure that the **developer** instructions in the README are still accurate and include at least minimal guidance for how to build the embeddings db (especially any gotchas).

Also, any chance we can make a local llm the default for now? I think it's important that Charlie evaluates how things are working with the local llm.. given that many users may be working with sensitive data.

Not for this PR.. just thinking.. How difficult would it be to make an initial app splash screen that allows a user to select between a local or other llms (openai, claud, mistral, etc) and mention that an api key is needed if they select a non-local llm.

ahuang11 commented:
> Not for this PR.. just thinking.. How difficult would it be to make an initial app splash screen that allows a user to select between a local or other llms (openai, claud, mistral, etc) and mention that an api key is needed if they select a non-local llm.

Maybe don't need a splash screen--part of the ChatInterface to `prompt_user` on init
https://github.com/holoviz/panel/pull/7148

droumis commented:
> Maybe don't need a splash screen--part of the ChatInterface to prompt_user on init

Yea that could work

droumis commented:
Thoughts re: "Also, any chance we can make a local llm the default for now? I think it's important that Charlie evaluates how things are working with the local llm.. given that many users may be working with sensitive data." ??

ahuang11 commented:
Yep, I've set the LLM to default to llamacpp:
```python
ui = lmai.ExplorerUI(
    agents=[lmai.agents.ChatAgent(tools=[doc_lookup], template_overrides={"main": {"instructions": INSTRUCTIONS}})],
    llm=lmai.llm.LlamaCpp(),
    default_agents=[], log_level="debug"
)
```

ahuang11 commented:
Let me add an entrypoint though to panel serve the app.py

ahuang11 commented:
Okay ready!

droumis commented:
do you have a `.pre-commit-config.yaml` that you can include?

droumis commented on README.md:
```diff
@@ -49,6 +51,7 @@ For a simple setup use [`uv`](https://docs.astral.sh/uv/):
 uv venv
 source .venv/bin/activate # on linux. Similar commands for windows and osx
 uv pip install -e .[dev]
+uv pip install "git+https://github.com/holoviz/lumen@main#egg=lumen[ai]"
```
```suggestion
uv pip install "git+https://github.com/holoviz/lumen@main#egg=lumen[ai-llama]"
```

droumis commented on README.md:
```diff
@@ -49,6 +51,7 @@ For a simple setup use [`uv`](https://docs.astral.sh/uv/):
 source .venv/bin/activate # on linux. Similar commands for windows and osx
 uv pip install -e .[dev]
+uv pip install "git+https://github.com/holoviz/lumen@main#egg=lumen[ai]"
 pre-commit run install
```
```suggestion
pre-commit run
```

droumis commented:
Had to manually run `CMAKE_ARGS="-DLLAMA_METAL=on -DCMAKE_OSX_ARCHITECTURES=arm64" FORCE_CMAKE="1" pip install 'llama-cpp-python[server]'`

Hitting an error when I ask anything

Maybe running out of memory with llama? Or maybe the install wasn't good

<details> <summary> Code </summary>

```bash
lumen-anndata
2025-05-12 13:19:31,554 Starting Bokeh server version 3.7.3 (running on Tornado 6.4.2)
2025-05-12 13:19:31,554 User authentication hooks NOT provided (default user enabled)
2025-05-12 13:19:31,557 Bokeh app running at: http://localhost:5006/app
2025-05-12 13:19:31,557 Starting Bokeh server with process id: 15968
2025-05-12 13:19:33,129 __________________________________________________________
2025-05-12 13:19:33,129 New Session: Started
2025-05-12 13:19:34,033 WebSocket connection opened
2025-05-12 13:19:34,035 ServerConnection created
ggml_metal_init: skipping kernel_get_rows_bf16                     (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32                   (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32_1row              (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32_l4                (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_bf16                  (not supported)
ggml_metal_init: skipping kernel_mul_mv_id_bf16_f32                (not supported)
ggml_metal_init: skipping kernel_mul_mm_bf16_f32                   (not supported)
ggml_metal_init: skipping kernel_mul_mm_id_bf16_f32                (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h64           (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h80           (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h96           (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h112          (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h128          (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h192          (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_hk192_hv128   (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h256          (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_hk576_hv512   (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_h96       (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_h128      (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_h192      (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_hk192_hv128 (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_h256      (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_hk576_hv512 (not supported)
ggml_metal_init: skipping kernel_cpy_f32_bf16                      (not supported)
ggml_metal_init: skipping kernel_cpy_bf16_f32                      (not supported)
ggml_metal_init: skipping kernel_cpy_bf16_bf16                     (not supported)
2025-05-12 13:19:50,269 __________________________________________________________
2025-05-12 13:19:50,269 New Message: 'how do scanpy output get stored in anndata'
ggml_metal_init: skipping kernel_get_rows_bf16                     (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32                   (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32_1row              (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32_l4                (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_bf16                  (not supported)
ggml_metal_init: skipping kernel_mul_mv_id_bf16_f32                (not supported)
ggml_metal_init: skipping kernel_mul_mm_bf16_f32                   (not supported)
ggml_metal_init: skipping kernel_mul_mm_id_bf16_f32                (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h64           (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h80           (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h96           (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h112          (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h128          (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h192          (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_hk192_hv128   (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h256          (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_hk576_hv512   (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_h96       (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_h128      (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_h192      (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_hk192_hv128 (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_h256      (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_hk576_hv512 (not supported)
ggml_metal_init: skipping kernel_cpy_f32_bf16                      (not supported)
ggml_metal_init: skipping kernel_cpy_bf16_f32                      (not supported)
ggml_metal_init: skipping kernel_cpy_bf16_bf16                     (not supported)
2025-05-12 13:19:52,313 Planner00272.prompts['main']['template']:
The current date time is May 12, 2025 01:19 PM
You are the team lead responsible for creating a step-by-step plan to address user queries by assigning subtasks to specialized actors (agents and tools).

Ground Rules:
- Respect dependency chains: assign tasks only when input requirements are met
- Leverage existing memory instead of regenerating information if possible
- Stay within scope of the user's request (don't plot unless asked, etc.)
- Do not use the same actor twice in a row
- Plan only rather than executing (avoid details)
- It's often unnecessary to use the same actor multiple times in a single plan
- Never mention a lack of data sources or tables in your plan - assume your actors will handle data discovery
- Do not ignore the actor's exclusions and conditions.
- Tools require actor interpretation - always follow-up tools with agents

Below are the available actors. Respect the conditions and exclusions.

Provides: Outputs this actor adds to memory
Requires: Inputs this actor needs from memory or others
Conditions: When to use this actor
Not with: agents that can't be used alongside this expert

You already have these keys satisfied in memory:
- `sources`
- `outputs`
- `agent_tool_contexts`


🛠️ Tools:
- TableLookup: Discovers relevant tables using vector search, providing context for other agents.
  Requires: `sources`
  Provides: `vector_metaset`
  Conditions:
  - Best paired with ChatAgent for general conversation about data
  - Skip if sufficient context already exists in memory
  - Not useful for data related queries
  Not with:
  - IterativeTableLookup

- IterativeTableLookup: Looks up the most relevant tables and provides SQL schemas of those tables.
  Requires: `sources`
  Provides: `vector_metaset`, `sql_metaset`
  Conditions:
  - Skip if sufficient context already exists in memory
  - Useful for answering data queries
  - Avoid for follow-up questions when existing data is sufficient
  - Use only when existing information is insufficient for the current query
  Not with:
  - TableLookup


🧑‍💼 Agents:
- ChatAgent: Engages in conversations about high-level data topics, offering suggestions or insights to get started.
  Requires: ``
  Provides: ``
  Conditions:
  - Best for high-level information about data or general conversation
  - Can be used to describe available tables
  - Not useful for answering data specific questions
  - Must be paired with TableLookup or DocumentLookup




# Planning guidelines:

Build your plan in dependency order - ensure each actor's requirements are met by previous steps and
ensure you don't use incompatible actors together.
2025-05-12 13:19:52,314 Characters: 2679
ggml_metal_init: skipping kernel_get_rows_bf16                     (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32                   (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32_1row              (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_f32_l4                (not supported)
ggml_metal_init: skipping kernel_mul_mv_bf16_bf16                  (not supported)
ggml_metal_init: skipping kernel_mul_mv_id_bf16_f32                (not supported)
ggml_metal_init: skipping kernel_mul_mm_bf16_f32                   (not supported)
ggml_metal_init: skipping kernel_mul_mm_id_bf16_f32                (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h64           (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h80           (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h96           (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h112          (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h128          (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h192          (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_hk192_hv128   (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_h256          (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_bf16_hk576_hv512   (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_h96       (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_h128      (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_h192      (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_hk192_hv128 (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_h256      (not supported)
ggml_metal_init: skipping kernel_flash_attn_ext_vec_bf16_hk576_hv512 (not supported)
ggml_metal_init: skipping kernel_cpy_f32_bf16                      (not supported)
ggml_metal_init: skipping kernel_cpy_bf16_f32                      (not supported)
ggml_metal_init: skipping kernel_cpy_bf16_bf16                     (not supported)
Traceback (most recent call last):
  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.13/site-packages/lumen/ai/llm.py", line 237, in stream
    async for chunk in chunks:
                       ^^^^^^
TypeError: 'async for' requires an object with __aiter__ method, got generator

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.13/site-packages/lumen/ai/coordinator.py", line 1043, in _compute_execution_graph
    plan = await self._make_plan(
           ^^^^^^^^^^^^^^^^^^^^^^
    ...<2 lines>...
    )
    ^
  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.13/site-packages/lumen/ai/coordinator.py", line 886, in _make_plan
    async for reasoning in self.llm.stream(
    ...<9 lines>...
            previous_plans.append(reasoning.chain_of_thought)
  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.13/site-packages/lumen/ai/llm.py", line 245, in stream
    for chunk in chunks:
                 ^^^^^^
  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.13/site-packages/instructor/dsl/partial.py", line 167, in from_streaming_response
    yield from cls.model_from_chunks(json_chunks, **kwargs)
  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.13/site-packages/instructor/dsl/partial.py", line 232, in model_from_chunks
    for chunk in json_chunks:
                 ^^^^^^^^^^^
  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.13/site-packages/instructor/dsl/partial.py", line 274, in extract_json
    for chunk in completion:
                 ^^^^^^^^^^
  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.13/site-packages/llama_cpp/llama.py", line 2057, in <genexpr>
    return (ChatCompletionChunk(**chunk) for chunk in self.create_chat_completion(*args, **kwargs))  # type: ignore
                                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.13/site-packages/llama_cpp/llama_chat_format.py", line 314, in _convert_text_completion_chunks_to_chat
    for i, chunk in enumerate(chunks):
                    ~~~~~~~~~^^^^^^^^
  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.13/site-packages/llama_cpp/llama.py", line 1320, in _create_completion
    for token in self.generate(
                 ~~~~~~~~~~~~~^
        prompt_tokens,
        ^^^^^^^^^^^^^^
    ...<14 lines>...
        grammar=grammar,
        ^^^^^^^^^^^^^^^^
    ):
    ^
  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.13/site-packages/llama_cpp/llama.py", line 912, in generate
    self.eval(tokens)
    ~~~~~~~~~^^^^^^^^
  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.13/site-packages/llama_cpp/llama.py", line 646, in eval
    self._ctx.decode(self._batch)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "/Users/droumis/src/lumen-anndata/.venv/lib/python3.13/site-packages/llama_cpp/_internals.py", line 311, in decode
    raise RuntimeError(f"llama_decode returned {return_code}")
RuntimeError: llama_decode returned -3
```

</details>

ahuang11 commented:
Works for me, but I have a M2 Pro; what version do you have?
I have
> Name: llama_cpp_python Version: 0.3.4

droumis commented:
❯ pip list | grep llama_cpp_python
llama_cpp_python              0.2.24

droumis commented:
OK, then no need to hold up this PR if it's just an issue for my system.. we can track it in that issue I posted.

droumis commented:
Upgraded to 0.3.4 and it's working for me now, thanks

droumis commented:
Great work :)
